### PR TITLE
#[UIC-1812] Raise Jira Tickets and Navigate to dashboard

### PR DESCRIPTION
**Pull request by arpit-agarwal**
### CATEGORY

Choose one

- [ ] Bug Fix
- [x] Enhancement (new features, refinement)
- [ ] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY
Raise Jira Tickets and Navigate to dashboard.
You need to add restActions in chart meta data

- It supports OPEN a url via get method only. No additional header can be passed. You can optionally pass target to changes default opening in new tab/window
- You can execute any rest post api call provided you are authorised to do so by 3rd party App
- Sample data available at [rest-actions/rest-actions-format.json](https://github.com/Guavus/incubator-superset/blob/feature/UIC-1812-raise-ticket/rest-actions/rest-actions-format.json)

**Two API verifed and supported are below**

1. Open a dashboard. You can pass full url or relative to https://host:port/ . It will not append prefix automatically
1. Raise a JIRA ticket
     1. You need to create a post action as specifed above
     1. [Obtain a API](https://confluence.atlassian.com/cloud/api-tokens-938839638.html) key from JIRA.
    1. Create a Authorisation header
          1. Build a string of the form username:api-token. i.e `arpit.agarwal@guavus.com:<api-token>`
          1. Encode the string to Base64.  Online encoder [here](https://www.base64encode.net/)
          1. Supply an authorization header with format Authorization: Basic {encoded-string}. Make sure to replace {encoded-string} with your encoded string from Sub Step b.
  1. Pass header as specified in sample structure of rest actions
   1. Build the post payload template. sample payload template for Jira is documented in sample structure shared. For more details follow [this page](https://developer.atlassian.com/server/jira/platform/jira-rest-api-examples/)
 1. Only POST request are supported
 1. Optionally provide a template for `success_mesage` as per sample

**Feature Set**

- You can pass placeholder in post data template `%placeholder_name%`
- Exhaustive list of placeholder is as below
    - time
    - dashboard_title
    - dashboard_url
    - chart_title
    - chart_selection
    - username (logged in rvf user)
    - chart_query
- Reporter will always be the user whose API token is used
- Indicator on dashboard for Success/Failure/In_progress only for POST Actions
- label of rest action will be used as title in 3-dot menu
- InProgress and failed restAction will show an icon on respective chart with default tooltip `${action.lable} is in progress` or  `${action.lable} has failed`
- Successful rest action will pop a toast message with default message `${action.lable} was successful`
- `success_message` can also be temptalized using placholder `%placeholder%`
    - Placeholder can be properties of action object like label, url
    - Placeholder can also be properties of success reponse such as `key` from sample response of JIRA

         ``` 
          {"id":"255293","key":"MD-324","self":"https://guavus-jira.atlassian.net/rest/api/2/issue/255293"}
        ```
   

**Under development**

- [x] Placeholder for %username% 
- [x] Placeholder for %chart_selection% 
- [x] Indicator on dashboard Success/Failure/In_progress only for POST Actions
- [x] Cleanup and console.log removal
- [x] Integrate with knox
- [x] formatted Jira description

**Future work**

- UI for app-developer to configure JIRA and Drill to dashboard action
- Action should be invoked from a click of a cell/point of chart
- Post payload form for user input
- Updated reporter to logged-in superset user if feasible
- Increase Unit test coverage

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
Created Jira
-  https://guavus-jira.atlassian.net/browse/MD-311
- https://guavus-jira.atlassian.net/browse/MD-310

### TEST PLAN
Test e2e flow of these 2 actions
### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by bipinsoniguavus** [comment]
👍 for well documented summary!

File: superset/assets/src/chart/chartAction.js [hunk 1]
```diff
@@ -83,6 +84,30 @@ export const ANNOTATION_QUERY_FAILED = 'ANNOTATION_QUERY_FAILED';
...
+export function restActionSuccess(queryResponse, key) {
+  return { type: REST_ACTION_SUCCESS, queryResponse, key };
+}
+
+export const REST_ACTION_STARTED = 'REST_ACTION_STARTED';
+export function restActionStarted(queryController,restAction, key) {
+  return { type: REST_ACTION_STARTED, restAction, queryController, key };
+}
+
+export const REST_ACTION_FAILED = 'REST_ACTION__FAILED';
```
Comment (bipinsoniguavus) [inline superset/assets/src/chart/chartAction.js]: Seems like an extra `_` in `REST_ACTION__FAILED`

File: superset/assets/src/dashboard/containers/Chart.jsx [hunk 2]
```diff
@@ -36,6 +38,7 @@ function mapStateToProps(
     charts: chartQueries,
     dashboardInfo,
     dashboardState,
+    dashboardLayout: undoableLayout,
```
Comment (bipinsoniguavus) [inline superset/assets/src/dashboard/containers/Chart.jsx]: What is `undoableLayout`?

File: superset/assets/src/utils/restActions.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
+export function createPostPayload(payloadformat, data) {
+  let payload = JSON.parse(JSON.stringify(payloadformat)); // deepclone hack
+  replacePlaceholder(payload, data)
+  return payload;
+}
+
+const replacePlaceholder = (obj, data) => {
+  Object.keys(obj).forEach(key => {
+    let value = obj[key]
+    if (isString(value)) {
```
Comment (bipinsoniguavus) [inline superset/assets/src/utils/restActions.js]: It would be great if you can use `_.isString` from lodash.

File: superset/views/core.py [hunk 4]
```diff
@@ -836,6 +836,16 @@ def shortner(self):
 class Superset(BaseSupersetView):
     """The base views for Superset!"""
 
+    @expose('/rest_actions', methods=['POST'])
+    def restActions(self):
```
Comment (bipinsoniguavus) [inline superset/views/core.py]: we follow `_` separated function names, please modify as `rest_actions`to make it consistent.

Comment (bipinsoniguavus) [reply superset/views/core.py]: Also, shouldn't be the name as `execute_rest_action`?

File: superset/assets/src/dashboard/containers/Chart.jsx (hunk 2 above)
Comment (arpit-agarwal) [reply superset/assets/src/dashboard/containers/Chart.jsx]: I have used the same name as used by Superset at [this location](https://github.com/Guavus/incubator-superset/blob/master/superset/assets/src/dashboard/containers/DashboardHeader.jsx#L54) to be consistent

File: superset/assets/src/chart/chartAction.js (hunk 1 above)
Comment (arpit-agarwal) [reply superset/assets/src/chart/chartAction.js]: indeed. done

File: superset/views/core.py (hunk 4 above)
Comment (arpit-agarwal) [reply superset/views/core.py]: done

File: superset/assets/src/utils/restActions.js (hunk 3 above)
Comment (arpit-agarwal) [reply superset/assets/src/utils/restActions.js]: done

File: superset/assets/src/chart/chartAction.js [hunk 5]
```diff
@@ -164,6 +188,34 @@ export const ADD_CHART = 'ADD_CHART';
 export function addChart(chart, key) {
   return { type: ADD_CHART, chart, key };
 }
+export const RUN_REST_QUERY = 'RUN_REST_QUERY';
+export function runRestQuery(action,timeout = 60, key) {
+  return (dispatch) => {
+    if ("get" === (action.method).toLowerCase()) {
+      window.open(action.url, action.target || "_blank");
```
[the remark this answers is not in the document]
Comment (arpit-agarwal) [reply superset/assets/src/chart/chartAction.js]: As per current scope we will support only get request that can be opend in a new/existing `window`

File: superset/assets/src/dashboard/components/SliceHeader.jsx [hunk 6]
```diff
@@ -34,10 +34,12 @@ const propTypes = {
   updateSliceName: PropTypes.func,
   toggleExpandSlice: PropTypes.func,
   forceRefresh: PropTypes.func,
+  executeRestAction: PropTypes.func,
   exploreChart: PropTypes.func,
   exportCSV: PropTypes.func,
   canExportCSV: PropTypes.bool,
   editMode: PropTypes.bool,
+  currentRestAction: PropTypes.object,
   annotationQuery: PropTypes.object,
```
[the remark this answers is not in the document]
Comment (arpit-agarwal) [reply superset/assets/src/dashboard/components/SliceHeader.jsx]: i used `current` to distinguish it from `restActions`. This is the action currently executing

**Comment by bipinsoniguavus** [comment]
Two issues which I have observed:

1. Overriding the chart will let loose the info of JIRA json
2. Before the charts gets it data, if I trigger action of raising JIRA tickets, it will fetch the old filters.

**Comment by arpit-agarwal** [comment]
@bipinsoniguavus 
1. has been fixed
2.  we will raise as a minor bug as it is not the postive usecase

**Comment by arpit-agarwal** [comment]
Test and lint fixed

**Pull request merged**